library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,53 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - Every type is ABI-compatible with itself.
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`, `NonNull<T>` are all ABI-compatible with each
+///   other for all `T`.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
+/// - Any two 1-ZST types (types with size 0 and alignment 1) are ABI-compatible.
```

Comment:
im not sure i love the term "1-ZST type" perhaps "ZST with alignment 1" as i think "type" is redundant here and the "1-" is a bit implicit for someone scanning through the docs. another option would be to link to some docs where the nature of 1-ZSTs are explained in more detail but im not sure that exists/where to put it.